text/0000-let-else.md

```diff
@@ -0,0 +1,652 @@
+
+## Desugaring example
+
+```rust
+let Some(x) = y else { return; };
+```
+
+Desugars to
+
+```rust
+let x = match y {
+    Some(x) => x,
+    _ => {
+        let nope: ! = { return; };
+        match nope {}
+    }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## The diverging block
+
+"Must diverge" is an unusual requirement, which doesn't exist elsewhere in the language as of the time of writing, 
+and might be difficult to explain or lead to confusing errors for programmers new to this feature.
+
+However, rustc does have support for representing the divergence through the type-checker via `!` or any other uninhabited type,
+so the implementation is not a problem.
+
+## `let PATTERN = if {} else {} else {};`
+
+One unfortunate combination of this feature with regular if-else expressions is the possibility of `let PATTERN = if { a } else { b } else { c };`.
+This is likely to be unclear if anyone writes it, but does not pose a syntactical issue, as `let PATTERN = if y { a } else { b };` should always be
+interpreted as `let Enum(x) = (if y { a } else { b });` (still a compile error as there no diverging block: `error[E0005]: refutable pattern in local binding: ...`)
+because the compiler won't interpret it as `let PATTERN = (if y { a }) else { b };` since `()` is not an enum.
+
+This can be overcome by making a raw if-else in the expression position a compile error and instead requiring that parentheses are inserted to disambiguate:
+`let PATTERN = (if { a } else { b }) else { c };`.
+
+Rust already provides us with such a restriction, and so the expression can be restricted to be a [`ExpressionWithoutBlock`][expressions].
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+let-else attempts to be as consistent as possible to similar existing syntax.
+
+Fundamentally it is treated as a `let` statement, necessitating an assignment and the trailing semicolon.
+
+Pattern matching works identically to if-let, no new "negation" pattern matching rules are introduced.
+
+Operator precedence with `&&` in made to be like if-let, requiring that a case which is an error prior to this RFC be changed to be a slightly different error.
+This is for a possible extension for let-else similar to the (yet unimplemented) if-else-chains feature, as mentioned in [future-possibilities][] with more detail.
+Specifically, while the following example is an error today, by the default `&&` operator rules it would cause problems with if-let-chains like `&&` chaining:
+
+```rust
+let a = false;
+let b = false;
+
+// The RFC proposes boolean patterns with a lazy boolean operator (&& or ||)
+//  be made into a compile error, for potential future compatibility with if-let-chains.
+let true = a && b else {
+    return;
+};
+```
+
+The expression can be any [`ExpressionWithoutBlock`][expressions], in order to prevent `else {} else {}` confusion, as noted in [drawbacks][#drawbacks].
```

Comment:
```suggestion
The expression can be any expression not ending in a brace, in order to prevent `else {} else {}` confusion, as noted in [drawbacks][#drawbacks].
```